ownCloud/Client/Actions/Action.swift

```diff
@@ -0,0 +1,218 @@
+	static let moreFolder: OCExtensionLocationIdentifier = OCExtensionLocationIdentifier("moreFolder") //!< Present in "more" options for a whole folder
+	static let toolbar: OCExtensionLocationIdentifier = OCExtensionLocationIdentifier("toolbar") //!< Present in a toolbar
+}
+
+class ActionExtension: OCExtension {
+	// MARK: - Custom Instance Properties.
+	var name: String
+	var category: ActionCategory
+
+	// MARK: - Init & Deinit
+	init(name: String, category: ActionCategory = .normal, identifier: OCExtensionIdentifier, locations: [OCExtensionLocationIdentifier]?, features: [String : Any]?, objectProvider: OCExtensionObjectProvider?, customMatcher: OCExtensionCustomContextMatcher?) {
+
+		self.name = name
+		self.category = category
+
+		super.init(identifier: identifier, type: .action, locations: locations, features: features, objectProvider: objectProvider, customMatcher: customMatcher)
+	}
+}
+
+class ActionContext: OCExtensionContext {
+	// MARK: - Custom Instance Properties.
+	weak var viewController: UIViewController?
+	weak var core: OCCore?
+	weak var query: OCQuery?
+	var items: [OCItem]
+
+	// MARK: - Init & Deinit.
+	init(viewController: UIViewController, core: OCCore, query: OCQuery? = nil, items: [OCItem], location: OCExtensionLocation, requirements: [String : Any]? = nil, preferences: [String : Any]? = nil) {
+		self.items = items
+
+		super.init()
+
+		self.viewController = viewController
+		self.core = core
+		self.location = location
+
+		self.query = query
+		self.requirements = requirements
+		self.preferences = preferences
+	}
+}
+
+class Action : NSObject {
+	// MARK: - Extension metadata
+	class var identifier : OCExtensionIdentifier? { return nil }
+	class var category : ActionCategory? { return .normal }
+	class var name : String? { return nil }
+	class var locations : [OCExtensionLocationIdentifier]? { return nil }
+	class var features : [String : Any]? { return nil }
+
+	// MARK: - Extension creation
+	class var actionExtension : ActionExtension {
+		let objectProvider : OCExtensionObjectProvider = { (_ rawExtension, _ context, _ error) -> Any? in
+			if let actionExtension = rawExtension as? ActionExtension,
+				let actionContext   = context as? ActionContext {
+				return self.init(for: actionExtension, with: actionContext)
+			}
+
+			return nil
+		}
+
+		let customMatcher : OCExtensionCustomContextMatcher  = { (context, priority) -> OCExtensionPriority in
+
+			guard let actionContext = context as? ActionContext else {
+				return priority
+			}
+
+			if self.applicablePosition(forContext: actionContext) == .none {
+				// Exclude actions whose applicablePosition returns .none
+				return .noMatch
+			}
+
+			if actionContext.items[0].type == OCItemType.collection, identifier!.rawValue == "com.owncloud.action.openin" {
+				return .noMatch
+			}
+
+			let priority = OCExtensionPriority(rawValue: priority.rawValue + UInt(self.applicablePosition(forContext:actionContext).rawValue))!
+			return priority
```

Comment:
I think you can just return `priority` here as `Action.sortedApplicableActions` will take care of sorting the actions according to their `applicablePosition`.

Reply:
Done, I wasn't using `Action.sortedApplicableActions`, now I am.

Reply:
Awesome, thanks!